migrate.go

```diff
@@ -117,6 +129,42 @@ func collectMigrations(dirpath string, current, target int64) (mm *MigrationMap,
     return mm, nil
 }
 
+// collect all migrations that have not been run, including ones older than the current version
+// This is only valid for migrating up
+func collectAllUnappliedMigrations(dirpath string, current, target int64, db *sql.DB) (mm *MigrationMap, err error) {
+    mm = &MigrationMap{}
+
+    // extract the numeric component of each migration,
+    // filter out any uninteresting files,
+    // and ensure we only have one file per migration version.
+    filepath.Walk(dirpath, func(name string, info os.FileInfo, err error) error {
+
+        if v, e := numericComponent(name); e == nil {
+
+            for _, m := range mm.Migrations {
+                if v == m.Version {
+                    log.Fatalf("more than one file specifies the migration for version %d (%s and %s)",
+                        v, m.Source, filepath.Join(dirpath, name))
+                }
+            }
+
+            row := db.QueryRow("SELECT version_id, is_applied from goose_db_version where version_id=$1;", v)
+
+            var record MigrationRecord
+            err := row.Scan(&record.VersionId, &record.IsApplied)
+            if err == sql.ErrNoRows {
+                mm.Append(v, name)
+            } else if versionFilter(v, current, target) {
```

Comment:
I don't think this check belongs here - the else-if should probably be based on the record.IsApplied value instead